File: FloodFill.py
# Time Complexity : O(nxm)
# Space Complexity : O(nxm)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : NA

# Approach is to fill the new color if its not original color using depth first traversal method by starting from given index.

class Solution:

    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        orig = image[sr][sc]
        if(image == None or len(image) ==0):
            return 
        if(orig == color):
            return image
        self.dfs(image, sr,sc, orig, color)
        return image

    def dfs(self, image, r, c , orig, color):
        if(r<0 or r==len(image) or c<0 or c==len(image[0]) or image[r][c]!= orig):
            return
        image[r][c]=color
        dirs=[[-1,0], [0,-1], [1,0],[0,1]]
        for dir in dirs:
            nr=r+dir[0]
            nc=c+dir[1]
            self.dfs(image, nr,nc, orig, color)

File: FloodFill.cs
// Time Complexity : O(MN)
// Space Complexity : O(MN)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


public class Solution {
    public int[][] FloodFill(int[][] image, int sr, int sc, int color) {
        if(image == null || image.Length == 0) return image;
        var m  = image.Length; var n = image[0].Length;
        var dirs = new List<(int, int)> { (0,1), (0, -1), (-1, 0), (1,0)};
        if(image[sr][sc] == color) return image;
        int oldcolor = image[sr][sc];
        dfs(image, sr, sc, m, n, oldcolor, color, dirs);
        return image;
    }
    private void dfs(int[][] image, int sr, int sc, int m, int n, 
            int oldcolor, int color, List<(int, int)> dirs)
    {
        if(sr < 0 || sc < 0 || sc == n || sr == m || image[sr][sc] != oldcolor) return;

        image[sr][sc] = color;
        foreach(var dir in dirs)
        {
            var r = sr + dir.Item1;
            var c = sc + dir.Item2;
            dfs(image, r,c,m,n,oldcolor, color, dirs);
        }
    }
}
